packit/config/common_package_config.py

```
@@ -596,3 +596,20 @@ def get_package_names_as_env(self) -> dict[str, str]:
                 return env
             raise PackitConfigException("No packages in config")
         raise PackitConfigException("Multiple packages in config")
+
+    def get_specfile_path_env(self) -> dict[str, str]:
+        """Creates a dict with a single key for specfile_path,
+        to be set in the environment variables
+
+        If the specfile_path value is not set in the
+        config file, it defaults to <downstream_package_name>.spec
+        :param action: The current action being carried out, used for logging info
+        :return: dict[str, str]
+        """
+
+        if not (specfile_path := self.specfile_path):
+            specfile_path = f"{self.downstream_package_name}.spec"
```

This part is more tricky because it behaves differently for upstream and downstream actions. See https://github.com/packit/packit/blob/a1ed5bc5bb6b4e590da5934213a8c3e252f8cfb2/packit/config/common_package_config.py#L483

I would say that for now it should just throw an error if `specfile_path` is not defined and revisit to see how this needs to be refactored to account for the different action types

---

> Optional: rename the `get_specfile_path_env` to something like `get_base_env`/`get_base_action_env`/etc. and move the the definition to the top of the `env` definitions i.e. it should define `env` and all other steps are manipulating it.

Okay gotten, my next question would be can we merge this get_base_env() method with get_package_names_as_env(), since that defines env variables that every action references, and my intention is to make the specfile path available to all actions. I thought about it at first but I didn't want to change too many things in the current implementation.

---

> This part is more tricky because it behaves differently for upstream and downstream actions. See
> 
> https://github.com/packit/packit/blob/a1ed5bc5bb6b4e590da5934213a8c3e252f8cfb2/packit/config/common_package_config.py#L483
> 
> I would say that for now it should just throw an error if `specfile_path` is not defined and revisit to see how this needs to be refactored to account for the different action types


Hmmm I don't quite understand, I thought there was a defined behavior already for this ie the path specified in the config file is the upstream path and it defaults to the downstream package path?

---

> can we merge this get_base_env() method with get_package_names_as_env()

Can't tell at first glance. Need to look more carefully at the code at the PC.

---

> Hmmm I don't quite understand, I thought there was a defined behavior already for this ie the path specified in the config file is the upstream path and it defaults to the downstream package path?

I believe I have an open issue about the expansion not being reliable. Don't know if @majamassarini or @mfocko remember or can dig up the issue for that.

But for this specific case it's more about reimplementing the behaviour in another place there. If there was a function that would expand to one or the other via that logic then you should use that. You could abstract that part, although you should probably address the issue that I'm trying to remember first, or there might be some clashing changes

---

It's clear now, thank you for providing context. To the extent that I reviewed the codebase, there's only one fallback value for the specfile which is the downstream_package.spec. The issue you highlighted would still linger with that fallback implementation tho